url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/protocols/specifications/YKHLProtocol.java

```diff
@@ -0,0 +1,149 @@
+package protocols.specifications;
+
+import java.math.BigDecimal;
+
+public class YKHLProtocol extends DBProtocol{
+	
+	private static final long serialVersionUID = -7060835761171756392L;
+	protected double pd;
+	protected int sizeOfNonce;
+	
+	public YKHLProtocol(double pd){
+		this(pd, SIZE_OF_NONCES);
+	}
+	
+	public YKHLProtocol(double pd, int sizeOfNonces){
+		this.pd = pd;
+		this.sizeOfNonce = sizeOfNonces;
+	}
+	
+	
+	public YKHLProtocol() {
+	}
+
+	@Override
+	public String getAcronym() {
+		return "YKHL";
+	}
+
+	@Override
+	public BigDecimal getMafiaFraudProbability() {
+		if (pd == 1) return mafiaFraudPdOne(n);
+		else if (pd == 0.75) return mafiaFraudPdThreeQuarter(n);
+		else throw new UnsupportedOperationException();
+	}
+
+	@Override
+	public DBProtocol getInstance() {
+		return new YKHLProtocol(pd);
+	}
+
+	private BigDecimal mafiaFraudPdOne(int n) {
+		BigDecimal tmp = new BigDecimal(""+((double)5)/8);
+		tmp = tmp.pow(n);
+		for (int i = 1; i <= n; i++) {
+			BigDecimal tmp2 = new BigDecimal(""+((double)1)/4);
+			BigDecimal tmp3 = new BigDecimal(""+((double)5)/8);
+			BigDecimal tmp4 = new BigDecimal(""+((double)3)/4);
+			tmp3 = tmp3.pow(i-1);
+			tmp4 = tmp4.pow(n-i);
+			tmp = tmp.add(tmp2.multiply(tmp3).multiply(tmp4));
+		}
+		return tmp;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 42:52, 57:67.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.